text/0000-variance.md

```diff
@@ -0,0 +1,568 @@
+  fact completely unnecessary. Defaulting to a particular variance would
+  not help in identifying these cases (though a better dead code lint might).
+- There is no default that is always correct but invariance, and
+  invariance is typically too strong.
+- Phantom type parameters occur relatively rarely anyhow.
+  
+**Remove variance inference and use fully explicit declarations.**
+Variance inference is a rare case where we do non-local inference
+across type declarations. It might seem more consistent to use
+explicit declarations. However, variance declarations are notoriously
+hard for people to understand. We were unable to come up with a
+suitable set of keywords or other system that felt sufficiently
+lightweight. Moreover, explicit annotations are error-prone when
+compared to the phantom data and fn approach (see example in the
+section regarding marker traits).
+
+# Unresolved questions
+
+There is one significant unresolved question: the correct way to
+handle a `*mut` pointer. It was revealed recently that while the
+current treatment of `*mut T` is correct, it frequently yields overly
+conservative inference results in practice. At present the inference
+treats `*mut T` as invariant with respect to `T`: this is correct and
+sound, because a `*mut` represents aliasable, mutable data, and indeed
+the subtyping relation for `*mut T` is that `*mut T <: *mut U if T=U`.
+
+However, in practice, `*mut` pointers are often used to build safe
+abstractions, the APIs of which do not in fact permit aliased
+mutation. Examples are `Vec`, `Rc`, `HashMap`, and so forth. In all of
+these cases, the correct variance is covariant -- but because of the
+conservative treatment of `*mut`, all of these types are being
+inferred to an invariant result.
+
+The complete solution to this seems to have two parts. First, for
+convenience and abstraction, we should not be building safe
+abstractions on raw `*mut` pointers anyway. We should have several
+convenient newtypes in the standard library, like `ptr::Unique`, that
+can be used, which would also help for handling OIBIT conditions and
+`NonZero` optimizations. In my branch I have used the existing (but
+unstable) type `ptr::Unique` for the primary role, which is kind of an
+"unsafe box". `Unique` should ensure that it is covariant with respect
+to its argument.
+
+However, this raises the question of how to implement `Unique` under
+the hood, and what to do with `*mut T` in general. There are various
+options:
+
+1. Change `*mut` so that it behaves like `*const`. This unfortunately
+   means that abstractions that introduce shared mutability have
+   a responsibility for add phantom data to that affect, something
+   like `PhantomData<*const Cell<T>>`. This seems non-obvious and
+   unnatural.
+   
+2. Rewrite safe abstractions to use `*const` (or even `usize`) instead
+   of `*mut`, casting to `*mut` only they have a `&mut self`
+   method. This is probably the most conservative option.
```

Comment:
Yes, I have currently implemented option 2: that is, `*mut` is invariant. I had been leaning towards what I think was 3 (ignore `*mut` references entirely and reauire explicit markers), but I fear that that is more error-prone. Basically if people aren't thinking carefully it's easy to make a covariant type that ought to have been invariant. Option 2 is correct by default, and requires you to opt-in (by using distinct types) to covariance. On the variance branch I've just been rewriting standard library stuff not to use raw `*mut` pointers but rather safe wrappers like `Unique`, which I think improves the code overall, since it's more declarative. It also handles the OIBIT requirements and so forth. Those are currently unstable though, we'll probably want some iteraiton on their design -- right now they are just kind of shallow wrappers around a raw pointer, we might be able to do better. (Note also that `Unique` on the branch is quite different from `Unique` on `master`).